### PR TITLE
Less restricted client version checking

### PR DESCRIPTION
**Pull request by frsyuki**
This is how it works:

* digdag-cli sends its version to the server (unless `--disable-version-check`
  is set)

* If server returns `apiCompatible: false`, client aborts. Client must
  upgrade.
  This is useful when there are too old clients and they may cause
  unexpected problems due to incompatible REST API version. But most
  likely, digdag shouldn't introduce such incompatible REST API changes.

* If server returns `upgradeRecommended: true`

  * If STDOUT is TTY, client aborts. Client must upgrade.
    This is useful when system administrator of the server wants to
    force all users upgrade to make technical support eassy.

  * Otherwise, show warning and continue processing.
    This is necessary when digdag command is run by a script where
    the command shouldn't abort.

Server returns `apiCompatible: false` when client version is older than
`server.client-version-check.api-incompatible-if-older` property.

Server returns `upgradeRecommended: true` when client version is older than
`server.client-version-check.upgrade-recommended-if-older` property.

### CODE REVIEW

**Comment by frsyuki** [comment]
@danielnorberg can I ask you to take a look?
About REST API, I think `/api/version/check` is OK but don't know it's the best or not. Please advice if you have ideas.

File: digdag-server/src/main/java/io/digdag/server/ClientVersionChecker.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+    private ClientVersionChecker(
+            Optional<Version> upgradeRecommendedUntil,
+            Optional<Version> apiIncompatibleUntil)
+    {
+        this.upgradeRecommendedUntil = upgradeRecommendedUntil;
+        this.apiIncompatibleUntil = apiIncompatibleUntil;
+    }
+
+    public boolean isUpgradeRecommended(Version clientVesrion)
```
Comment (danielnorberg) [inline digdag-server/src/main/java/io/digdag/server/ClientVersionChecker.java]: s/clientVesrion/clientVersion

File: digdag-server/src/main/java/io/digdag/server/ClientVersionChecker.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    {
+        if (!upgradeRecommendedUntil.isPresent()) {
+            return false;
+        }
+        return clientVesrion.isOlder(upgradeRecommendedUntil.get());
+    }
+
+    public boolean isApiCompatible(Version clientVesrion)
```
Comment (danielnorberg) [inline digdag-server/src/main/java/io/digdag/server/ClientVersionChecker.java]: s/clientVesrion/clientVersion

File: digdag-tests/src/test/java/acceptance/VersionIT.java [hunk 3]
```diff
@@ -41,31 +46,73 @@ public void testVersionHeader()
...
+        assertThat(interactive.code(), is(0));  // success
+
+        // batch succeeds
+        CommandStatus batch = main(LocalVersion.of(localVersion).withBatchModeCheck(true), "workflows", "-e", server.endpoint(), "-c", config.toString());
+        assertThat(batch.code(), is(0));  // success
     }
 
     @Test
-    public void verifyVersionMismatchRejected()
+    public void testVersionCheckWithUpgradeRecommendeVersion()
```
Comment (danielnorberg) [inline digdag-tests/src/test/java/acceptance/VersionIT.java]: s/testVersionCheckWithUpgradeRecommendeVersion/testVersionCheckWithUpgradeRecommendedVersion

File: digdag-tests/src/test/java/acceptance/VersionIT.java [hunk 4]
```diff
@@ -41,31 +46,73 @@ public void testVersionHeader()
...
+        assertThat(interactive.errUtf8(), containsString("server: " + REMOTE_VERSION));
+        assertThat(interactive.errUtf8(), containsString(
+                    "This client version is obsoleted. It is recommended to upgrade"));
+        assertThat(interactive.errUtf8(), containsString(
+                    "Please run following command locally to upgrade to the latest version compatible to the server:"));
+        assertThat(interactive.errUtf8(), containsString("digdag selfupdate "  + REMOTE_VERSION));
+
+        // batch mode warns
+        CommandStatus batch = main(LocalVersion.of(localVersion).withBatchModeCheck(true), "workflows", "-e", server.endpoint(), "-c", config.toString());
+        assertThat(batch.code(), is(0));  // success
```
Comment (danielnorberg) [inline digdag-tests/src/test/java/acceptance/VersionIT.java]: Check warning output?

**Comment by danielnorberg** [comment]
Looks good apart from a few spelling errors 👍

File: digdag-tests/src/test/java/acceptance/VersionIT.java (hunk 4 above)
Comment (frsyuki) [inline digdag-tests/src/test/java/acceptance/VersionIT.java]: That is better but I couldn't find good ways to do that...I tried to hook appender but it didn't work somehow. Do you have good ideas to hook logs?

```diff
diff --git a/digdag-tests/src/test/java/acceptance/VersionIT.java b/digdag-tests/src/test/java/acceptance/VersionIT.java
index d6a3247..27e0e55 100644
--- a/digdag-tests/src/test/java/acceptance/VersionIT.java
+++ b/digdag-tests/src/test/java/acceptance/VersionIT.java
@@ -1,6 +1,7 @@
 package acceptance;
 
 import io.digdag.client.Version;
+import io.digdag.cli.client.ClientCommand;
 import org.junit.Before;
 import org.junit.Rule;
 import org.junit.Test;
@@ -12,6 +13,7 @@ import utils.TemporaryDigdagServer;
 import java.nio.file.Files;
 import java.nio.file.Path;
 
+import static utils.TestUtils.hookLogger;
 import static utils.TestUtils.main;
 import static org.hamcrest.Matchers.containsString;
 import static org.hamcrest.Matchers.is;
@@ -83,8 +85,17 @@ public class VersionIT
         assertThat(interactive.errUtf8(), containsString("digdag selfupdate "  + REMOTE_VERSION));
 
         // batch mode warns
-        CommandStatus batch = main(LocalVersion.of(localVersion).withBatchModeCheck(true), "workflows", "-e", server.endpoint(), "-c", config.toString());
-        assertThat(batch.code(), is(0));  // success
+        String logs = hookLogger(ClientCommand.class, () -> {
+            CommandStatus batch = main(LocalVersion.of(localVersion).withBatchModeCheck(true), "workflows", "-e", server.endpoint(), "-c", config.toString());
+            assertThat(batch.code(), is(0));  // success
+        });
+        assertThat(logs, containsString("client: " + localVersion));
+        assertThat(logs, containsString("server: " + REMOTE_VERSION));
+        assertThat(logs, containsString(
+                    "This client version is obsoleted. This client is subject to be incompatible and stop working in near future"));
+        assertThat(logs, containsString(
+                    "Please run following command locally to upgrade to the latest version compatible to the server:"));
+        assertThat(logs, containsString("digdag selfupdate "  + REMOTE_VERSION));
     }
 
     @Test
diff --git a/digdag-tests/src/test/java/utils/TestUtils.java b/digdag-tests/src/test/java/utils/TestUtils.java
index 7a5599b..209fd01 100644
--- a/digdag-tests/src/test/java/utils/TestUtils.java
+++ b/digdag-tests/src/test/java/utils/TestUtils.java
@@ -1,5 +1,10 @@
 package utils;
 
+import ch.qos.logback.classic.spi.ILoggingEvent;
+import ch.qos.logback.core.Appender;
+import ch.qos.logback.core.OutputStreamAppender;
+import ch.qos.logback.core.UnsynchronizedAppenderBase;
+import ch.qos.logback.core.layout.EchoLayout;
 import com.amazonaws.AmazonServiceException;
 import com.amazonaws.services.s3.AmazonS3;
 import com.amazonaws.services.s3.model.DeleteObjectsRequest;
@@ -732,4 +737,69 @@ public class TestUtils
             }
         }
     }
+
+    public interface ThrowingRunnable <E extends Exception>
+    {
+        void run() throws E;
+    }
+
+    public static <E extends Exception> String hookLogger(Class<?> loggerName, ThrowingRunnable<E> hooked)
+            throws E
+    {
+        ch.qos.logback.classic.Logger logger = (ch.qos.logback.classic.Logger) LoggerFactory.getLogger(loggerName);
+
+        final StringBuilder sb = new StringBuilder();
+        //ByteArrayOutputStream out = new ByteArrayOutputStream();
+
+        logger.error("hooking logger appender: "+logger.isAdditive());
+        logger.setAdditive(true);
+
+        System.out.println("appenders:");
+        Iterable<Appender<ILoggingEvent>> ite = () -> ((ch.qos.logback.classic.Logger) logger).iteratorForAppenders();
+        for (Appender<ILoggingEvent> app : ite) {
+            System.out.println("appender : " + app + " class="+app.getClass());
+        }
+
+        UnsynchronizedAppenderBase<ILoggingEvent> appender = new UnsynchronizedAppenderBase<ILoggingEvent>() {
+            @Override
+            public void append(ILoggingEvent event)
+            {
+                System.out.println("event: " + event);
+                System.err.println("event: " + event);
+
+                sb.append(event.getMessage()).append("\n");
+            }
+
+            @Override
+            public void doAppend(ILoggingEvent event)
+            {
+                System.out.println("event: " + event);
+                System.err.println("event: " + event);
+
+                sb.append(event.getMessage()).append("\n");
+            }
+        };
+        //appender.setOutputStream(out);
+        //appender.setLayout(new EchoLayout<>());
+        appender.setName("digdagHookForTest");
+        appender.start();
+
+        System.out.println("hooking logger: " + System.identityHashCode(logger) + " class="+logger.getClass());
+
+        ((ch.qos.logback.classic.Logger) logger).addAppender(appender);
+        try {
+            System.out.println("start..");
+            hooked.run();
+        }
+        finally {
+            ((ch.qos.logback.classic.Logger) logger).detachAppender("digdagHookForTest");
+            appender.stop();
+        }
+
+        logger.error("hooking logger appender");
+        logger.error("done: " + sb.toString());
+
+        return sb.toString();
+        //return new String(out.toByteArray(), UTF_8);
+    }
 }
```

File: digdag-server/src/main/java/io/digdag/server/ClientVersionChecker.java (hunk 1 above)
Comment (frsyuki) [inline digdag-server/src/main/java/io/digdag/server/ClientVersionChecker.java]: 👍

File: digdag-server/src/main/java/io/digdag/server/ClientVersionChecker.java (hunk 2 above)
Comment (frsyuki) [inline digdag-server/src/main/java/io/digdag/server/ClientVersionChecker.java]: 👍

File: digdag-tests/src/test/java/acceptance/VersionIT.java (hunk 3 above)
Comment (frsyuki) [inline digdag-tests/src/test/java/acceptance/VersionIT.java]: 👍

**Comment by frsyuki** [comment]
@danielnorberg Thank you!

**Pull request merged**